### PR TITLE
update lately npu modification

### PR DESCRIPTION
**Pull request by ckirchhoff2021**
## Motivation

This PR contains some npu-adapters of mmcv ops, which contain nms, focal_loss, masked_conv and modulated_deform_conv. we can run these ops(above) on ascend-npu device by these adapters. 

## Modification

nms_npu.cpp is added to the PR which is used for nms npu-adaptation, while focal_loss_npu.cpp, masked_conv,py, modulated_deform_conv.py and test_masked_conv.py are modified by the PR. we have fixed some kernel bugs and modify the focal_loss adapter to improve the op performance. masked_conv and modulated_deform_conv can be conveniently adapted by some torch_npu.ops, as a result , we can make them available for npu by just modifying the py scripts. Finally, we modify the ut test for masked_conv2d which can make it running on npu-device.

## Q & A 
welcome to contact us for more information if you have any questions about the PR.

### CODE REVIEW

File: .pre-commit-config-zh-cn.yaml [hunk 1]
```diff
@@ -62,11 +62,11 @@ repos:
...
-  #       language: system
-  #       files: \.(c|cc|cxx|cpp|cu|h|hpp|hxx|cuh|proto)$
+  - repo: local
+    hooks:
+      - id: clang-format
+        name: clang-format
+        description: Format files with ClangFormat
+        entry: clang-format -style=google -i
+        language: system
+        files: \.(c|cc|cxx|cpp|cu|h|hpp|hxx|cuh|proto)$
```
Comment (zhouzaida) [inline .pre-commit-config-zh-cn.yaml:72]: The hook should not be enabled.

File: mmcv/ops/masked_conv.py [hunk 2]
```diff
@@ -45,6 +45,24 @@ def forward(ctx,
...
+                padding=(pad_h, pad_w),
+                dilation=(1, 1),
+                groups=1)
+            if mask.size()[1:] != conv.size()[2:]:
+                raise ValueError(
+                    'The mask is inconsistent with the shape of output_conv.')
+            conv_h, conv_w = conv.size()[2:]
+            mask_reshape = mask.reshape(1, 1, conv_h, conv_w)
+            output = conv * mask_reshape
+            return output
```
Comment (zhouzaida) [inline mmcv/ops/masked_conv.py:64]:
```suggestion
        if features.device.type == 'npu':
            import torch_npu
            output = torch_npu.npu_conv2d(
                features,
                weight,
                bias,
                stride=(stride_h, stride_w),
                padding=(pad_h, pad_w),
                dilation=(1, 1),
                groups=1)
            if mask.size()[1:] != output.size()[2:]:
                raise ValueError(
                    'The mask is inconsistent with the shape of output_conv.')
            out_h, out_w = output.size()[2:]
            mask_reshape = mask.reshape(1, 1, out_h, out_w)
            output = output * mask_reshape
            return output
```

File: mmcv/ops/masked_conv.py [hunk 3]
```diff
@@ -45,6 +45,24 @@ def forward(ctx,
...
+                bias,
+                stride=(stride_h, stride_w),
+                padding=(pad_h, pad_w),
+                dilation=(1, 1),
+                groups=1)
+            if mask.size()[1:] != conv.size()[2:]:
+                raise ValueError(
+                    'The mask is inconsistent with the shape of output_conv.')
+            conv_h, conv_w = conv.size()[2:]
+            mask_reshape = mask.reshape(1, 1, conv_h, conv_w)
```
Comment (zhouzaida) [inline mmcv/ops/masked_conv.py:62]: Is it necessary to reshape the mask to (1, 1, conv_h, conv_w)?

File: mmcv/ops/modulated_deform_conv.py [hunk 4]
```diff
@@ -34,6 +34,72 @@ def symbolic(g, input, offset, mask, weight, bias, stride, padding,
...
+        ]
+        sort_index_for_npu_fp = torch.IntTensor(sort_index_for_npu_fp)
+        sort_index_for_npu_bp = torch.IntTensor(sort_index_for_npu_bp)
+        sort_index_for_npu_fp = sort_index_for_npu_fp.npu()
+        sort_index_for_npu_bp = sort_index_for_npu_bp.npu()
+        return sort_index_for_npu_fp, sort_index_for_npu_bp
+
+    @staticmethod
+    def _npu_forward(ctx, input_tensor, offset, mask, weight, bias):
+        _, _, k1, k2 = weight.shape
```
Comment (zhouzaida) [inline mmcv/ops/modulated_deform_conv.py:58]:
```suggestion
        _, _, kernel_h, kernel_w = weight.shape
```

File: mmcv/ops/modulated_deform_conv.py [hunk 5]
```diff
@@ -34,6 +34,72 @@ def symbolic(g, input, offset, mask, weight, bias, stride, padding,
...
+            ModulatedDeformConv2dFunction._calculate_npu_sort_index(
+                k2, k1, ctx.deform_groups)
+        select_offset = offset.index_select(1, sort_index_for_npu_fp)
+        offset_all = torch.cat([select_offset, mask], dim=1)
+        output, offset_out = torch.npu_deformable_conv2d(
+            input_tensor,
+            weight,
+            offset_all,
+            conv2d_bias,
+            kernel_size=[k2, k1],
```
Comment (zhouzaida) [inline mmcv/ops/modulated_deform_conv.py:70]: What is the order of these two parameters, (h, w) or (w, h)?

File: mmcv/ops/modulated_deform_conv.py [hunk 6]
```diff
@@ -34,6 +34,72 @@ def symbolic(g, input, offset, mask, weight, bias, stride, padding,
             groups_i=groups,
             deform_groups_i=deform_groups)
 
+    @staticmethod
+    def _calculate_npu_sort_index(kernel_h, kernel_w, deformable_group):
+        split_num = deformable_group * 2 * kernel_h * kernel_w
+        sort_index_for_npu = list(range(split_num))
```
Comment (zhouzaida) [inline mmcv/ops/modulated_deform_conv.py:40]: Suggest removing `_npu` in variables.

**Review by zhouzaida — APPROVED** [review]
